### PR TITLE
Initial .gitlab-ci.yml build file

### PR DESCRIPTION
**Pull request by ddorgan**

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    - ./.cargo/
+
+
+.collect_artifacts:                &collect_artifacts
+  artifacts:
+    name:                          "${CI_JOB_NAME}_${CI_COMMIT_REF_NAME}"
+    when:                          on_success
+    expire_in:                     7 days
+    paths:
+    - target/release/polkadot
```
Comment (sorpaas) [inline .gitlab-ci.yml]: I think this should be `target/release/shasper`.

Comment (ddorgan) [reply .gitlab-ci.yml]: Woops sorry about that...fixed.

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -0,0 +1,64 @@
+stages:
+  - test
+  - build
+
+image:                             parity/rust:nightly
```
Comment (sorpaas) [inline .gitlab-ci.yml]: Not sure the implication of `nightly` tag here. However, for `rust`, stable version should still be the default -- we only use `nightly` to build the wasm runtime, and build everything else on `stable`.

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+
+cache:
+  key:                             "${CI_JOB_NAME}"
+  paths:
+    - ./target/
+    - ./.cargo/
+
+
+.collect_artifacts:                &collect_artifacts
```
Comment (sorpaas) [inline .gitlab-ci.yml]: Do we also need `.releasable_branches` configs if we enabled `.collect_artifacts`?

File: .gitlab-ci.yml (hunk 2 above)
Comment (ddorgan) [reply .gitlab-ci.yml]: This image should have both actually. The reason to keep both inside is that we avoid cross-link errors when attempting to do a rustup on an existing docker image.

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -0,0 +1,64 @@
+stages:
+  - test
+  - build
+
+image:                             parity/rust:nightly
+
+variables:
+  CI_SERVER_NAME:                  "GitLab CI"
+  CARGO_HOME:                      "${CI_PROJECT_DIR}/.cargo"
+
```
Comment (sorpaas) [inline .gitlab-ci.yml]: Not sure how this works here, but do we need to set `BUILD_TARGET`, `BUILD_ARCH`, `CARGO_TARGET` envs here?

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+before_script:
+  - ./scripts/init.sh
+  - export PATH="${CI_PROJECT_DIR}/cargo/bin/:$PATH"
+  - ./scripts/build.sh
+
+
+#### stage:                        test
+
+test:rust:stable:
```
Comment (sorpaas) [inline .gitlab-ci.yml]: Do we need to set `RUST_TOOLCHAIN` configs here (which should be `stable`)?

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  stage:                           test
+  script:
+    - time cargo test --all --locked
```
Comment (sorpaas) [inline .gitlab-ci.yml]: Sorry about the confusion on this, but I rethink about this and we probably still need `--release` parameters here. The issue is that when we get any integration tests, debug build tests might start to be slow.

File: .gitlab-ci.yml [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  tags:
+    - rust-stable
+
+
+#### stage:                        build
+
+build:linux:ubuntu:amd64:
+  stage:                           build
+  script:
+    - cargo build
```
Comment (sorpaas) [inline .gitlab-ci.yml]: If we're collecting artifacts here, I think this should be `--release`!

File: .gitlab-ci.yml (hunk 5 above)
Comment (5chdn) [reply .gitlab-ci.yml]: please name it `test-rust-stable`, `:` will cause issues if we run this on windows in future

File: .gitlab-ci.yml [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    - master
+    - schedules
+    - web
+  tags:
+    - rust-stable
+
+
+#### stage:                        build
+
+build:linux:ubuntu:amd64:
```
Comment (5chdn) [inline .gitlab-ci.yml]: `build-linux` is sufficient, unless we have more complex ci setup here

**Pull request merged**